### PR TITLE
NativeAOT-LLVM: Remove ILLink target

### PR DESCRIPTION
**Pull request by yowl**
This PR removes the `dependson` for `PrepareForILLink` and the `_RunILLink` .

I think this will fix: #2235

### CODE REVIEW

**Comment by yowl** [comment]
cc @dotnet/nativeaot-llvm

File: src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets [hunk 1]
```diff
@@ -118,7 +118,7 @@ The .NET Foundation licenses this file to you under the MIT license.
     <IlcCompileDependsOn Condition="'$(BuildingFrameworkLibrary)' != 'true'">Compile;ComputeIlcCompileInputs</IlcCompileDependsOn>
     <IlcCompileDependsOn Condition="'$(IlcMultiModule)' == 'true' and '$(BuildingFrameworkLibrary)' != 'true'">$(IlcCompileDependsOn);BuildFrameworkLib</IlcCompileDependsOn>
     <IlcCompileDependsOn>$(IlcCompileDependsOn);SetupOSSpecificProps</IlcCompileDependsOn>
-    <IlcCompileDependsOn>$(IlcCompileDependsOn);PrepareForILLink</IlcCompileDependsOn>
+    <IlcCompileDependsOn Condition="'$(NativeCodeGen)' != 'llvm'">$(IlcCompileDependsOn);PrepareForILLink</IlcCompileDependsOn>
```
Comment (SingleAccretion) [inline src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: What is the original reason for the dependence on `PrepareForILLink`?

Comment (yowl) [reply src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets]: I think its an extension point for developers to hook in their own steps before Ilc runs.  Not great to remove it in that case.  I'll have another go at this.

**Comment by yowl** [comment]
How about this, I've just added an empty `PrepareForILLink` target so that users can add before and after targets?  I think this is all we get out of this task for Naot.  @MichalStrehovsky would know better what we are losing, if anything with an empty target.

**Comment by yowl** [comment]
From what I read, PrepareForILLink was added as part of a commit to make NAOT similar in behaviour to PublishTrimmed.

**Comment by yowl** [comment]
The latest commit adds a call to `_ComputeManagedAssemblyForILLink` to get similar behaviour, I think for the `PublishTrimmed` modes (partial?).

**Comment by MichalStrehovsky** [comment]
PrepareForILLink set up everything for trimming (what assemblies should be trimmed, how, etc.):

https://github.com/dotnet/runtime/blob/2b530d4abdab90801b9c3434d62e511d9f72e3e7/src/tools/illink/src/ILLink.Tasks/build/Microsoft.NET.ILLink.targets#L172-L183

Why are we not including the trimming targets on WASM?

**Comment by yowl** [comment]
Upstream for (e.g win-x64) will call `PrepareForILLink` as a `DependsOn` for `WriteIlcRspFileForCompilation`  .  `PrepareForILLink` is defined in `Microsoft.NET.ILLink.targets` from the package `microsoft.net.illink.tasks` which does not get included for NativeAOT-LLVM (its done via the SDK, not sure?).

Maybe we should just make it a requirement (and document it), that the user has to add
```
   <PackageReference Include="microsoft.net.illink.tasks" Version="8.0.0-*" />
```
to their csproj, that seems to work?

**Comment by MichalStrehovsky** [comment]
> to their csproj, that seems to work?

Yep, that sounds better. Best would be to figure out why it's not getting included (are we setting PublishTrimmed too late?) but both are better than the empty target.

**Comment by yowl** [comment]
> Best would be to figure out why it's not getting included

Basically I start with the fact that `ILLinkTargetsPath` is not set for LLVM.  That leads me to the `console2.csproj.nuget.dgspec.json` file where for winx64 there is ILLink as a dependency
```json
"frameworks": {
  "net8.0": {
    "targetAlias": "net8.0",
    "dependencies": {
      "Microsoft.DotNet.ILCompiler": {
        "suppressParent": "All",
        "target": "Package",
        "version": "[8.0.0-preview.3.23174.8, )",
        "autoReferenced": true
      },
      "Microsoft.NET.ILLink.Tasks": {
        "suppressParent": "All",
        "target": "Package",
        "version": "[8.0.0-preview.3.23174.8, )",
        "autoReferenced": true
      }
    },
```

Where as for LLVM that section is
```json
  "frameworks": {
    "net8.0": {
      "targetAlias": "net8.0",
      "dependencies": {
        "Microsoft.DotNet.ILCompiler.LLVM": {
          "target": "Package",
          "version": "[8.0.0-*, )"
        },
        "runtime.win-x64.Microsoft.DotNet.ILCompiler.LLVM": {
          "target": "Package",
          "version": "[8.0.0-*, )"
        }
      },
```
how does `.nuget.dgspec.json` get generated ?

**Comment by MichalStrehovsky** [comment]
That would be question that has answers somewhere in the dotnet/sdk repo. Does it get referenced if you add `<PublishTrimmed>true</PublishTrimmed>` to the project?

**Comment by yowl** [comment]
> Does it get referenced if you add `<PublishTrimmed>true</PublishTrimmed>` to the project?

Yes it does, will look at dotnet/sdk to see what I can find.  Thanks.

**Comment by yowl** [comment]
So without `PublishAot==true` (https://github.com/dotnet/runtimelab/issues/2235#issuecomment-1481415589) this logic leaves 
`_IsTrimmingEnabled==false`

```
  <PropertyGroup>
    <DefaultCopyToPublishDirectoryMetadata Condition="'$(DefaultCopyToPublishDirectoryMetadata)' == ''">true</DefaultCopyToPublishDirectoryMetadata>
    <_GetChildProjectCopyToPublishDirectoryItems Condition="'$(_GetChildProjectCopyToPublishDirectoryItems)' == ''">true</_GetChildProjectCopyToPublishDirectoryItems>
    <IsPublishable Condition="'$(IsPublishable)' == ''">true</IsPublishable>
    <!-- PublishAot depends on PublishTrimmed. This must be set early enough for the KnownILLinkPack to be restored. -->
    <PublishTrimmed Condition="'$(PublishTrimmed)' == '' And '$(PublishAot)' == 'true'">true</PublishTrimmed>
    <_IsTrimmingEnabled Condition="'$(_IsTrimmingEnabled)' == '' And ('$(PublishTrimmed)' == 'true' Or '$(IsTrimmable)' == 'true')">true</_IsTrimmingEnabled>
    <_IsTrimmingEnabled Condition="'$(_IsTrimmingEnabled)' == ''">false</_IsTrimmingEnabled>
    <DynamicCodeSupport Condition="'$(DynamicCodeSupport)' == '' And '$(PublishAot)' == 'true'">false</DynamicCodeSupport>
  </PropertyGroup>
```

I'll go with `<PublishTrimmed>true</PublishTrimmed>`   as that is implied to be true for NativeAOT .

**Comment by yowl** [comment]
cc @dotnet/nativeaot-llvm 

I think this is good as we'll get it.  Have not tested .net 7 but I'm not sure we really care, I don't.

File: docs/using-nativeaot/compiling.md [hunk 2]
```diff
@@ -64,7 +64,17 @@ For WebAssembly, it is always a cross-architecture scenario as the compiler runs
 <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="8.0.0-*" />
 ```
 
-Then, the required package reference is
+Then, remove 
+```xml
+<NativeAot>true</NativeAot>
```
Comment (jkotas) [inline docs/using-nativeaot/compiling.md]: Should this be `<PublishAot>true</PublishAot>`?

Comment (yowl) [reply docs/using-nativeaot/compiling.md]: Yes, thanks.

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**